rosbag2_storage_mcap/src/mcap_storage.cpp

```diff
@@ -531,25 +558,76 @@ bool MCAPStorage::read_and_enqueue_message()
   return true;
 }
 
-namespace
-{
-bool is_service_event_topic_name(const std::string & topic_name)
+bool MCAPStorage::is_topic_name_a_service_event(const std::string_view topic_name)
 {
   // The origin definition is RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX
-  const char * service_introspection_topic_postfix = "/_service_event";
-  if (topic_name.length() <= strlen(service_introspection_topic_postfix)) {
+  static const char * service_event_topic_postfix = "/_service_event";
+  static const size_t service_event_topic_postfix_len = strlen(service_event_topic_postfix);
+  size_t topic_name_len = topic_name.length();
+  if (topic_name_len <= service_event_topic_postfix_len) {
     return false;
   }
-
-  std::string end_topic_name =
-    topic_name.substr(topic_name.length() - strlen(service_introspection_topic_postfix));
-  if (end_topic_name != service_introspection_topic_postfix) {
+  auto end_topic_name = topic_name.substr(topic_name_len - service_event_topic_postfix_len);
+  if (end_topic_name != service_event_topic_postfix) {
     return false;
   }
-
   return true;
 }
-}  // namespace
+
+template <typename T>
+bool MCAPStorage::is_topic_selected_by_white_list_or_regex(const std::string_view topic_name,
+                                                           const T & white_list,
+                                                           const std::string & regex)
+{
+  if (!white_list.empty()) {
+    if (std::find(white_list.begin(), white_list.end(), topic_name) != white_list.end()) {
+      return true;
+    } else if (!regex.empty()) {  // Topic not found in non-empty white_list and regex not empty
+      std::smatch m;
+      std::string topic_string(topic_name);
+      std::regex re(regex);
+      if (std::regex_match(topic_string, m, re)) {
+        return true;
+      } else {
+        // Topic name not found in non-empty regex
+        return false;
+      }
+    }
+  } else if (regex.empty()) {
+    // Both white list and regex are empty. i.e. equivalent to include all
+    return true;
+  } else {  // if (white_list.empty() && !regex.empty())
+    std::smatch m;
+    std::string topic_string(topic_name);
+    std::regex re(regex);
+    if (std::regex_match(topic_string, m, re)) {
+      return true;
+    } else {
+      // Topic name not found in non-empty regex
+      return false;
+    }
+  }
+  return false;
```

Comment:
There is duplicated code for regex.   
Can it be simplified like this? 

```suggestion
  // Both white list and regex are empty. i.e. equivalent to include all
  if (white_list.empty() && regex.empty()) {
    return true;
  }

  if (std::find(white_list.begin(), white_list.end(), topic_name) != white_list.end()) {
      return true;
  }
  
  if (!regex.empty()) {
    std::smatch m;
    std::string topic_string(topic_name);
    std::regex re(regex);
    if (std::regex_match(topic_string, m, re)) {
      return true;
    }
  }

  return false;
```